File: 2015/03/03b.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <utility>
#include <set>


int main (void) {

    std::string problem_input;
    std::fstream file;

    file.open("./input.txt", std::ios::in);

    if (file.is_open()) {
        while (getline(file, problem_input)) {}
        file.close();
    };

    std::pair<int, int> coord_santa (0, 0);
    std::pair<int, int> coord_robot (0, 0);

    std::vector<std::pair<int, int>> coordinates_list;
    coordinates_list.push_back(coord_santa);
    coordinates_list.push_back(coord_robot);

    for (int i = 0; i < problem_input.length(); i++) {
        
        std::pair<int, int>* coordinates;
        if (i % 2 != 0) {coordinates = &coord_santa;}
        else            {coordinates = &coord_robot;}

        switch (problem_input[i]) {
            case '<':
                coordinates->first  -= 1;
                break;
            case '>':
                coordinates->first  += 1;         
                break;
            case 'v':
                coordinates->second -= 1;
                break;
            case '^':
                coordinates->second += 1;
                break;
        }

        coordinates_list.push_back(*coordinates);

    }

    std::set<std::pair<int, int>> coordinates_set(coordinates_list.begin(), coordinates_list.end());
    coordinates_list.assign(coordinates_set.begin(), coordinates_set.end());

    std::cout << coordinates_set.size() << std::endl;

}

File: 2015/03/03b.py
with open('./input.txt', 'r') as file:
    problem_input = file.read()

coord_santa, coord_robot = [0, 0], [0, 0]
coordinates_list = [tuple(coord_santa), tuple(coord_robot)]

for sorter, direction in enumerate(problem_input, 1):

    coordinates = coord_santa if sorter % 2 != 0 else coord_robot

    match direction:
        case '<':
            coordinates[0] -= 1
        case '>':
            coordinates[0] += 1
        case 'v':
            coordinates[1] -= 1
        case '^':
            coordinates[1] += 1

    coordinates_list.append(tuple(coordinates))

print(len(set(coordinates_list)))